### PR TITLE
feat: conditional policies on flow steps

### PR DESCRIPTION
**Pull request by remisultan**
closes: https://github.com/gravitee-io/issues/issues/7016

## How to Test

1. Create a Policy flow (either at domain or application level) to allow a validation of Conditional policy (see below)

Screenshot 2022-03-08 at 22 05 07
Screenshot 2022-03-08 at 22 05 15
Screenshot 2022-03-08 at 22 05 28

Reverse the condition in the condition of the step to be able to assert the use

### CODE REVIEW

**Review by leleueri — CHANGES_REQUESTED** [review]

**Comment by leleueri** [comment]
When I want to update an existing EL into the condition fields, the changed are applied on the search text field after enter a space (probably more related to the UI component part) :

image

File: gravitee-am-gateway/gravitee-am-gateway-policy/src/main/java/io/gravitee/am/gateway/policy/impl/PolicyChain.java [hunk 1]
```diff
@@ -87,6 +90,25 @@ public void doNext(Request request, Response response) {
...
+        final String condition = policy.condition();
+        if (isNullOrEmpty(condition) || condition.isBlank()) {
+            return true;
+        }
+
+        final TemplateEngine templateEngine = executionContext.getTemplateEngine();
+        try {
+            return nonNull(templateEngine) && templateEngine.getValue(condition.trim(), Boolean.class);
+        } catch (Exception e) {
+            LOGGER.error("Could not execute rule [{}] for policy [{}]", condition, policy.id(), e);
```
[the remark this answers is not in the document]
Comment (remisultan) [reply gravitee-am-gateway/gravitee-am-gateway-policy/src/main/java/io/gravitee/am/gateway/policy/impl/PolicyChain.java:107]: I agree

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-common/src/main/java/io/gravitee/am/gateway/handler/common/flow/execution/ExecutionPolicy.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+import io.gravitee.am.gateway.policy.Policy;
+import java.util.Objects;
+
+/**
+ * @author Rémi SULTAN (remi.sultan at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+public class ExecutionPolicy {
+    private final Policy policy;
```
Comment (remisultan) [inline gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-common/src/main/java/io/gravitee/am/gateway/handler/common/flow/execution/ExecutionPolicy.java:27]: That file is useless

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-common/src/test/java/io/gravitee/am/gateway/handler/common/flow/FlowManagerTest.java [hunk 3]
```diff
@@ -79,7 +76,7 @@ public void shouldNoFindByExtensionPoint_noFlows() {
             Assert.assertTrue(policies.isEmpty());
             return true;
         });
-        verify(policyPluginManager, never()).create(anyString(), anyString());
+        verify(policyPluginManager, never()).create(anyString(), eq(null), anyString());
```
Comment (farmborough) [inline gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-common/src/test/java/io/gravitee/am/gateway/handler/common/flow/FlowManagerTest.java:79]: Might as well use `isNull()`.

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-common/src/test/java/io/gravitee/am/gateway/handler/common/vertx/web/handler/PolicyChainHandlerTest.java [hunk 4]
```diff
@@ -107,14 +107,14 @@ public void shouldInvoke_onePolicy() {
 
         policyChainHandler.handle(routingContext);
 
-        verify(flowManager, times(1)).findByExtensionPoint(eq(ExtensionPoint.PRE_CONSENT), eq(null), any(FlowPredicate.class));
+        verify(flowManager, times(1)).findByExtensionPoint(eq(ExtensionPoint.PRE_CONSENT), eq(null), any(ExecutionPredicate.class));
```
Comment (farmborough) [inline gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-common/src/test/java/io/gravitee/am/gateway/handler/common/vertx/web/handler/PolicyChainHandlerTest.java:110]: `isNull()`?

**Review by farmborough — APPROVED** [review]

**Review by tcompiegne — APPROVED** [review]

**Comment by tcompiegne** [comment]
Well done @remisultan 👍

**Pull request merged**